atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java

```diff
@@ -0,0 +1,319 @@
+    public void updateStateMachineForBucketAssigner(
+            BucketStateAndIdentifier original, BucketStateAndIdentifier updated) {
+        Cell cell = SweepAssignedBucketStoreKeyPersister.INSTANCE.sweepBucketAssignerStateMachineCell();
+        casCell(
+                cell,
+                Optional.of(bucketStateAndIdentifierPersister.trySerialize(original)),
+                bucketStateAndIdentifierPersister.trySerialize(updated));
+    }
+
+    @Override
+    public BucketStateAndIdentifier getBucketStateAndIdentifier() {
+        Cell cell = SweepAssignedBucketStoreKeyPersister.INSTANCE.sweepBucketAssignerStateMachineCell();
+        Optional<BucketStateAndIdentifier> value = readCell(cell, bucketStateAndIdentifierPersister::tryDeserialize);
+        return value.orElseThrow(() -> new SafeIllegalStateException(
+                "No bucket state and identifier found. This should have been bootstrapped during "
+                        + " initialisation, and as such, is an invalid state."));
+    }
+
+    @Override
+    public Set<Bucket> getStartingBucketsForShards(Set<ShardAndStrategy> shardAndStrategies) {
+        List<Cell> cellsToLoad = shardAndStrategies.stream()
+                .map(SweepAssignedBucketStoreKeyPersister.INSTANCE::sweepBucketPointerCell)
+                .collect(Collectors.toList());
+        Map<Cell, Value> result = keyValueService.getRows(
+                TABLE_REF,
+                cellsToLoad.stream().map(Cell::getRowName).collect(Collectors.toList()),
+                // TODO: This is janky.
+
+                ColumnSelection.create(
+                        cellsToLoad.stream().map(Cell::getColumnName).collect(Collectors.toSet())),
+                Long.MAX_VALUE);
+
+        return shardAndStrategies.stream()
+                .map(shardAndStrategy -> {
+                    Cell cell = SweepAssignedBucketStoreKeyPersister.INSTANCE.sweepBucketPointerCell(shardAndStrategy);
+                    return Optional.ofNullable(result.get(cell))
+                            .map(Value::getContents)
+                            .map(bucketIdentifierPersister::tryDeserialize)
+                            .map(bucketIdentifier -> Bucket.of(shardAndStrategy, bucketIdentifier))
+                            .orElseGet(() -> {
+                                log.warn(
+                                        "No starting bucket found for shard and strategy {}",
+                                        SafeArg.of("shardAndStrategy", shardAndStrategy));
+                                return Bucket.of(shardAndStrategy, 0);
+                            });
+                })
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    // TODO: See if we should centralise the work that was done in Jeremy's PR
+    public void updateStartingBucketForShardAndStrategy(Bucket newStartingBucket) {
```

Comment:
Thoughts on centralising this "bump upwards" logic vs duplicating it just once more here? This is a total of two usages, which isn't that many. pro: more specific log lines and naming, con: need to test again (but it's not that hard to write the tests)

Given it's only two usages, a reasonable option is to deduplicate if there's yet another need for this style of logic.